### PR TITLE
Add speed and speed perturbation functions and transforms

### PR DESCRIPTION
**Pull request by hwangjeff**
Adds functions and transforms for speed and speed perturbation (https://www.isca-speech.org/archive/interspeech_2015/ko15_interspeech.html).

### CODE REVIEW

File: torchaudio/prototype/transforms/_transforms.py [hunk 1]
```diff
@@ -384,3 +385,56 @@ def forward(self, waveform: torch.Tensor) -> torch.Tensor:
...
+
+
+def _source_target_sample_rate(orig_freq: int, speed: float) -> Tuple[int, int]:
+    source_sample_rate = int(speed * orig_freq)
+    target_sample_rate = int(orig_freq)
+    gcd = math.gcd(source_sample_rate, target_sample_rate)
+    return source_sample_rate // gcd, target_sample_rate // gcd
+
+
+class TimeStretch(torch.nn.Module):
```
Comment (mthrok) [inline torchaudio/prototype/transforms/_transforms.py:397]: what's your take on the name collision with the existing one?

https://pytorch.org/audio/stable/generated/torchaudio.transforms.TimeStretch.html#torchaudio.transforms.TimeStretch

Comment (hwangjeff) [reply torchaudio/prototype/transforms/_transforms.py:397]: ah that's unfortunate. i feel like "time stretch" fits better with the operators introduced here (stretching waveforms in time domain vs stretching spectrograms along time axis). this is also more consistent with librosa (https://librosa.org/doc/main/generated/librosa.effects.time_stretch.html)

Comment (hwangjeff) [reply torchaudio/prototype/transforms/_transforms.py:397]: actually, never mind — librosa's is neither: it accepts a waveform as input, generates spectrograms, and does the stretching on the spectrograms before converting back to time domain

Comment (hwangjeff) [reply torchaudio/prototype/transforms/_transforms.py:397]: let's go with "speed" — consistent with sox and "speed perturbation"

File: torchaudio/prototype/functional/functional.py [hunk 2]
```diff
@@ -295,3 +298,39 @@ def barkscale_fbanks(
...
+
+    Returns:
+        (torch.Tensor, torch.Tensor):
+            torch.Tensor
+                Speed-adjusted waveform, with shape `(..., new_time).`
+            torch.Tensor
+                Valid lengths of signals in speed-adjusted waveform, with shape `(...)`.
+    """
+
+    source_sample_rate = int(factor * orig_freq)
```
Comment (xiaohui-zhang) [inline torchaudio/prototype/functional/functional.py:327]: frequency and sample rate are the same thing, but maybe use terms more consistently. sample_rate -> freq as used in the implementation of "resample".  BTW are we flipping source and target here? I thought target_freq (or new_freq) should be int(factor * orig_freq) instead. Maybe I missed sth.

File: torchaudio/prototype/transforms/_transforms.py [hunk 3]
```diff
@@ -384,3 +385,100 @@ def forward(self, waveform: torch.Tensor) -> torch.Tensor:
         specgram = self.spectrogram(waveform)
         bark_specgram = self.bark_scale(specgram)
         return bark_specgram
+
+
+def _source_target_sample_rate(orig_freq: int, speed: float) -> Tuple[int, int]:
+    source_sample_rate = int(speed * orig_freq)
```
Comment (xiaohui-zhang) [inline torchaudio/prototype/transforms/_transforms.py:391]: same here

File: torchaudio/prototype/transforms/_transforms.py [hunk 4]
```diff
@@ -384,3 +385,100 @@ def forward(self, waveform: torch.Tensor) -> torch.Tensor:
...
+
+        Returns:
+            (torch.Tensor, torch.Tensor):
+                torch.Tensor
+                    Speed-adjusted waveform, with shape `(..., new_time).`
+                torch.Tensor
+                    Valid lengths of signals in speed-adjusted waveform, with shape `(...)`.
+        """
+
+        if torch.bernoulli(self.perturb_prob):
```
Comment (xiaohui-zhang) [inline torchaudio/prototype/transforms/_transforms.py:475]: it looks like all waveforms in a batch will be either all perturbed together or all kept original? can we apply the bernoulli factor on each waveform instead? that'll add more randomness in the data especially when batch size is large.

File: torchaudio/prototype/functional/functional.py (hunk 2 above)
Comment (xiaohui-zhang) [reply torchaudio/prototype/functional/functional.py:327]: ok thanks for the explanation. you're right

File: torchaudio/prototype/transforms/_transforms.py (hunk 4 above)
Comment (xiaohui-zhang) [reply torchaudio/prototype/transforms/_transforms.py:475]: ok I agree. I don't it worth the performance degradation. let's keep it the current way. BTW I think it can also be achieved by inserting "1"s in the "factors" sequence, e.g. if we want perturb 40% data, then we just make sure 60% of the elements in "factors" are "1"s and use perturb_prob=1.

**Review by xiaohui-zhang — APPROVED** [review]

File: torchaudio/prototype/transforms/_transforms.py (hunk 4 above)
Comment (hwangjeff) [reply torchaudio/prototype/transforms/_transforms.py:475]: that's a good idea. it wasn't the api that was originally proposed, but this seems cleaner and clearer
